File: vuln-sandbox/go/cwe-78/command-injection-safe.go
// SAFE: go-exec-command-bash-c — allowlist validation prevents command injection
// Rule: go-exec-command-bash-c | CWE-78 | Expected: TrueNegative

package main

import (
	"fmt"
	"net/http"
	"os/exec"
)

var allowedCommands = map[string]bool{
	"date":   true,
	"uptime": true,
	"pwd":    true,
}

func runHandler(w http.ResponseWriter, r *http.Request) {
	command := r.URL.Query().Get("cmd")

	// SAFE: command validated against an allowlist; exec.Command called with fixed args, no bash -c
	if !allowedCommands[command] {
		http.Error(w, "command not allowed", http.StatusBadRequest)
		return
	}

	cmd := exec.Command(command)
	out, err := cmd.Output()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	fmt.Fprintf(w, "%s", out)
}

func main() {
	http.HandleFunc("/run", runHandler)
	http.ListenAndServe(":8080", nil)
}

File: vuln-sandbox/typescript/cwe-78/command-injection-safe.ts
// SAFE: command-injection — allowlist validation and shell:false prevent injection (TypeScript)
// Rule: InjectChildProcessShellTrue | CWE-78 | Expected: TrueNegative

import express, { Request, Response } from 'express';
import { spawn } from 'child_process';

const app = express();
app.use(express.json());

const ALLOWED_COMMANDS: ReadonlySet<string> = new Set(['ls', 'pwd', 'date', 'uptime']);

app.get('/run', (req: Request, res: Response): void => {
  const command = req.query.cmd as string;

  // SAFE: command validated against allowlist; shell:false (default) prevents metacharacter injection
  if (!ALLOWED_COMMANDS.has(command)) {
    res.status(400).json({ error: 'Command not allowed' });
    return;
  }

  const proc = spawn(command, [], { shell: false });
  let output = '';
  proc.stdout.on('data', (data: Buffer) => { output += data.toString(); });
  proc.on('close', () => res.json({ output }));
});

app.listen(3000);
